src/routes/data.ts

```diff
@@ -0,0 +1,187 @@
+import { badGateway } from "@hapi/boom";
+import joi from "joi";
+import { PutObjectCommand, S3Client } from "@aws-sdk/client-s3";
+import { DataFormat, MIME_TYPE, StorageService } from "../constants.js";
+import { UnixFS } from "@web3-storage/upload-client";
+import type { Client as W3UpClient } from "@web3-storage/w3up-client";
+import type {
+    Lifecycle,
+    ReqRefDefaults,
+    ResponseToolkit,
+    ServerRoute,
+    Request,
+} from "@hapi/hapi";
+
+interface GetDataRoutesParams {
+    w3UpClient: W3UpClient;
+    s3Client: S3Client;
+    s3Bucket: string;
+}
+
+type DataFormatSerializer = (deserialized: any) => Promise<string> | string;
+
+type StorageServiceHandler = (
+    storageFormat: DataFormat,
+    request: Request<ReqRefDefaults>,
+    h: ResponseToolkit,
+) => Promise<Lifecycle.ReturnValue>;
+
+export const getDataRoutes = async ({
+    w3UpClient,
+    s3Client,
+    s3Bucket,
+}: GetDataRoutesParams): Promise<ServerRoute[]> => {
+    const SUPPORTED_FORMAT_SERIALIZERS: Record<
+        DataFormat,
+        DataFormatSerializer
+    > = {
+        [DataFormat.Json]: JSON.stringify,
+    };
+
+    const SUPPORTED_STORAGE_SERVICE_HANDLERS: Record<
+        StorageService,
+        StorageServiceHandler
+    > = {
+        [StorageService.Ipfs]: async (format, request, h) => {
+            const { data } = request.payload as { data: object };
+            const dataString = await SUPPORTED_FORMAT_SERIALIZERS[format](data);
+            const blob = new Blob([dataString], {
+                type: MIME_TYPE[format],
+            });
+
+            const precalculatedCid = (
+                await UnixFS.encodeFile(blob)
+            ).cid.toString();
+
+            let cidFromW3Up: string;
+            try {
+                cidFromW3Up = (await w3UpClient.uploadFile(blob)).toString();
+            } catch (error) {
+                request.logger.error(error, "Could not upload data to w3up");
+                return badGateway("Could not upload data to w3up");
+            }
+
+            if (cidFromW3Up !== precalculatedCid) {
+                return badGateway(
+                    `CID mismatch: expected ${precalculatedCid}, got ${cidFromW3Up}`,
+                );
+            }
+
+            request.logger.info(
+                `Data stored on ipfs with cid ${precalculatedCid} on behalf of ${request.auth.credentials.user}`,
+            );
+            return h
+                .response({ cid: precalculatedCid })
+                .code(200)
+                .type("application/json");
+        },
+        [StorageService.S3]: async (format, request, h) => {
+            const { data } = request.payload as { data: object };
+            const dataString = await SUPPORTED_FORMAT_SERIALIZERS[format](data);
```

Comment:
Same here for the `await`, unless we expect future serializers to return a promise for some reason.

Reply:
Same answer above applies here.